### PR TITLE
Require Node.js 12.20 and move to ESM

### PR DESCRIPTION
**Pull request by antongolub**
* add named exports
* update all deps to the latest versions
* add Node.js v16 to test matrix
* apply xo --fix
* BREAKING CHANGE: require Node.js >= 12.20
* BREAKING CHANGE: drop default exports, remove all globby.* methods

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this.

You need to update index.d.ts and the readme for ESM too.

File: index.js [hunk 1]
```diff
@@ -172,10 +167,21 @@ module.exports.stream = (patterns, options) => {
...
+export const hasMagic = (patterns, options) => [patterns].flat()
 	.some(pattern => fastGlob.isDynamicPattern(pattern, options));
 
-module.exports.gitignore = gitignore;
+export {
+	gitignore,
+	gitignoreSync,
+};
+
+// Legacy API
```
Comment (sindresorhus) [inline index.js:178]: This will be a major release. I don't want to expose the legacy API.

File: index.js [hunk 2]
```diff
@@ -172,10 +167,21 @@ module.exports.stream = (patterns, options) => {
...
+Object.assign(globby, {
+	hasMagic,
+	generateGlobTasks,
+	gitignore,
+	stream,
+	sync,
+});
+
+export default globby;
```
Comment (sindresorhus) [inline index.js:187]: There should not be a default export.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by antongolub** [comment]
Let's save `globby` as `globbyAsync` alias.

File: index.js [hunk 3]
```diff
@@ -107,18 +104,16 @@ const globToTask = task => glob => {
 
 	return {
 		pattern: glob,
-		options
+		options,
 	};
 };
 
-module.exports = async (patterns, options) => {
+export const globby = async (patterns, options) => {
```
[the remark this answers is not in the document]
Comment (antongolub) [reply index.js]: Let's save globby as globbyAsync alias.

File: bench.js [hunk 4]
```diff
@@ -1,95 +1,98 @@
...
 	process.chdir(BENCH_DIR);
-	['a', 'b']
-		.map(directory => `${directory}/`)
-		.forEach(directory => {
-			fs.mkdirSync(directory);
-			for (let i = 0; i < 500; i++) {
-				fs.writeFileSync(directory + (i < 100 ? 'c' : 'd') + i, '');
-			}
-		});
+	for (const directory of ['a', 'b']
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply bench.js:78]: Move it out into a variable to make it more readable.

**Comment by antongolub** [comment]
rfr

File: index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.js]: I disagree. Aliases cause confusion.

[the remark this answers is not in the document]
Comment (antongolub) [reply index.js]: Maybe just `globby` and `globbySync`?

**Comment by sindresorhus** [comment]
I can never decide how to handle named exports with async and sync methods:

- `globby` and `globbySync`
- `globbyAsync` and `globbySync`

The former looks nicer, but the latter is clearer. Any opinion?

**Comment by antongolub** [comment]
My guess is that for most users, the migration will look like this:
`const globby = require('globby') → import {globbyAsync as globby} from 'globby'`.
Well... I'd prefer just `import {globby} from 'globby'` instead.

**Comment by sindresorhus** [comment]
Alright. Let's go with `globby` for async and drop `globbyAsync`.

**Comment by antongolub** [comment]
@sindresorhus, your turn

**Comment by zloirock** [comment]
@sindresorhus sorry, why do you avoid the default export here?

**Comment by fisker** [comment]
I guess for consistency import? But I think maybe we can export both named and default? `export {globby as default, globby}`

**Comment by sindresorhus** [comment]
It's weird to have one default and one named when there are two main exports. I only do default and named export mix when there's only one main export and some secondary exports (like error, or helper utilities).

**Comment by sindresorhus** [comment]
> But I think maybe we can export both named and default? export {globby as default, globby}

No, there should be only one way to import the thing. Aliases create confusion and inconsistency.

**Pull request merged**